runtime/src/main/java/org/corfudb/runtime/view/replication/ChainReplicationProtocol.java

```diff
@@ -73,96 +73,140 @@ public ILogData peek(RuntimeLayout runtimeLayout, long globalAddress) {
+                .getLogUnitClient(startAddress, numUnits - 1)
+                .readRange(addresses)).getAddresses();
+
+        return readOrHoleFill(runtimeLayout, logResult, startAddress, numUnits, waitForWrite);
+    }
+
+    private Map<Long, ILogData> readOrHoleFill(RuntimeLayout runtimeLayout,
+                                               Map<Long, LogData> logResult,
+                                               long startAddress,
+                                               int numUnits,
+                                               boolean waitForWrite) {
+        Map<Long, ILogData> returnResult = new TreeMap<>(logResult);
+
+        // If waiting not required for writer, hole fill directly
+        if (!waitForWrite) {
+            logResult.forEach((address, value) -> {
+                if (!isLogDataPresent(value)) {
+                    holeFill(runtimeLayout, address);
+                    returnResult.put(address, peek(runtimeLayout, address));
+                }
+            });
+            return returnResult;
+        }
 
         // In case of holes, use the standard backoff policy for hole fill for
-        // the first entry in the list. All subsequent holes in the list can
-        // be hole filled without waiting as we have already waited for the first
-        // hole.
-        boolean wait = !waitForWrite;
-        Map<Long, ILogData> returnResult = new TreeMap<>();
+        // the first entry in the list. All subsequent holes in the list can be
+        // hole filled without waiting as we have already waited for the first hole.
+        boolean wait = true;
         for (Map.Entry<Long, LogData> entry : logResult.entrySet()) {
             long address = entry.getKey();
             ILogData value = entry.getValue();
-            if (value == null || value.isEmpty()) {
-                if (!wait) {
+            if (!isLogDataPresent(value)) {
+                if (wait) {
                     value = read(runtimeLayout, address);
-                    wait = true;
+                    wait = false;
                 } else {
                     // try to read the value
                     value = CFUtils.getUninterruptibly(runtimeLayout
                             .getLogUnitClient(startAddress, numUnits - 1).read(address))
                             .getAddresses().get(address);
                     // if value is null, fill the hole and get the value.
-                    if (value == null || value.isEmpty()) {
+                    if (!isLogDataPresent(value)) {
                         holeFill(runtimeLayout, address);
                         value = peek(runtimeLayout, address);
                     }
                 }
+                returnResult.put(address, value);
             }
-
-            returnResult.put(entry.getKey(), value);
         }
+
         return returnResult;
     }
 
+
+    private boolean isLogDataPresent(ILogData logData) {
```

Comment:
I dont think we need this. Just remove nulls from the lower layers.

Reply:
Done.